trinity/config.py

```diff
@@ -440,14 +452,60 @@ def nodekey(self, value: Union[bytes, PrivateKey]) -> None:
                 f"`PrivateKey` instance: got {type(self._nodekey)}"
             )
 
+    @contextmanager
+    def process_id_file(self, process_name: str):  # type: ignore
+        with PidFile(process_name, self.data_dir):
+            yield
+
     @classmethod
-    def from_parser_args(cls, parser_args: argparse.Namespace) -> 'TrinityConfig':
+    def from_parser_args(cls,
+                         parser_args: argparse.Namespace,
+                         sub_config_types: Iterable[Type['BaseSubConfig']]) -> 'BaseApplicationConfig':
         """
         Helper function for initializing from the namespace object produced by
         an ``argparse.ArgumentParser``
         """
         constructor_kwargs = construct_trinity_config_params(parser_args)
-        return cls(**constructor_kwargs)
+        trinity_config = cls(**constructor_kwargs)
+
+        trinity_config.initialize_sub_configs(parser_args, sub_config_types)
+
+        return trinity_config
+
+    def initialize_sub_configs(self,
+                               parser_args: argparse.Namespace,
+                               sub_config_types: Iterable[Type['BaseSubConfig']]) -> None:
+
+        for sub_config_type in sub_config_types:
+            self.add_sub_config(sub_config_type.from_parser_args(parser_args, self))
+
+    def add_sub_config(self, sub_config: 'BaseSubConfig') -> None:
+        self._sub_configs[type(sub_config)] = sub_config
+
+    def has_sub_config(self, sub_config_type: Type['BaseSubConfig']) -> bool:
+        return sub_config_type in self._sub_configs.keys()
+
+    def get_sub_config(self, sub_config_type: Type[TSubConfig]) -> TSubConfig:
+        # We want this API to return the specific type of the sub config that is requested.
+        # Our backing field only knows that it is holding `BaseSubConfig`'s but not concrete types
+        return cast(TSubConfig, self._sub_configs[sub_config_type])
+
+
+class BaseSubConfig:
+
+    base_config: BaseApplicationConfig = None
+    args: argparse.Namespace = None
+
+    @classmethod
+    @abstractmethod
+    def from_parser_args(cls, args: argparse.Namespace, base_config: BaseApplicationConfig) -> 'BaseSubConfig':
+        sub_config = cls()
```

Comment:
This feels weird to. So, now if you want define an alternative constructor (which is the thing we want to enable by introducing the `from_parser_args` API), then you would also have to also overwrite `from_parser_args` because the default one would just call the constructor without arguments before it sets the properties.

Not sure if you had something different in mind?

Reply:
I think that each `AppConfig` **must** implement their own `from_parser_args` and that this function body should just be a `pass` statement.  The individual app configs should follow the same structure as the `TrinityConfig` to convert the parser args into a set of constructor args and then construct the class.  In theory we can generalize this if we had an expectation for a `classmethod` or `staticmethod` that did the conversion from parser args to constructor args which would allow this method to be standardized.

Reply:
Ok, but even then we are still shifting the responsibility of where to store the `base_config` and `args` references into the sub classes. Unless, we do what I did in this PR and at least define `base_config` and `args` as properties as *suggestions*. But still, since each subclass would be free to just bypass this, there would be no guarantee that from every sub config you can access the base config at `base_config`.